students/Sally_Shen/lesson3/mailroom.py

```diff
@@ -0,0 +1,96 @@
+
+
+'''choms'''
+
+import sys
+
+print("in mailroom")
+'''Writing first name is easier for testing.'''
+donors ={"Fred Gates": [100.90, 200, 50], "Jeff Bazos": [100, 50, 600], "Mark Zuckerberg": [200, 50, 500],
+         "Paul Allen": [300, 40, 200]}
+
+
+def thank_you():
+    response = input("Type your donor name, if don't know, type list to pick")
+    '''If the user types list show them a list of the donor names and re-prompt'''
+    if response == "list":
+        for key in donors.keys():
+            print(key)
+        response = input("Please type your donor name.")
+
+    if response not in donors.keys():
+        donors[response] = []
+
+    newDonation = input("how much to donate?")
+    donors[response].append(float(newDonation))
+    email = '''
+    Dear {0},
+    Thank you for donating {1}!\n'''.format(response, newDonation)
+    print(email)
+
+
+def gen_stats(donor):
+    donations = donor[1]
+    total = sum(donations)
+    num  =  len(donations)
+    stats = (donor[0], total, num, total/num)
+    return
+
+def report():
+    amountToNames = {}
+    for name, amount in donors.items():
+        total = sum(amount)
+        if total in amountToNames.keys():
+            amountToNames[total].append(name)
+        else:
+            amountToNames[total] = [name]
+
+    sortedAmount = sorted(amountToNames, reverse=True)
+    output = "{:<15}${:<15}{:<12}${:<12}\n".format("Donor Name", "|Total Given", "|Num Gifts", "|Average Gift")
+    output += "-------------------------------------------------------------\n"
+
+
+    for amount in sortedAmount:
+        nameList = amountToNames[amount]
```

Comment:
Interesting -- I wouldn't have thought to do this way. But it should work, but a few issues:

There is a lot of coupling here -- a lot of this code will need to change if you change how the donors are stored. 

Better to extract all the donor info in one chunk, and work with that data structure.